### PR TITLE
PS-8050  PS 8.0 upgrade (from PS 5.7) crashes with Assertion failure:…

### PR DESCRIPTION
**Pull request by rahulmalik87**
… dict0dict.cc:1236:table2 == nullptr

https://jira.percona.com/browse/PS-8050

Problem:
crash while upgrade due to duplicate table space object.

Analysis:
During upgrade to 8.0, table_id are increased by DICT_MAX_DD_TABLES (1024). The
DICT_MAX_DD_TABLES are reserve for DD tables.
Subject 5.7 schema has table SYS_FIELDS with table_id 4 and SYS_ZIP_DICT
with table_id 1028. Once SYS_FIELDS is upgrade to 4 + 1024 = 1028.
PS asserts while trying to upgrade SYS_ZIP_DICT since server already has
table open with 1028

Fix:
SYS_* tables are special. All SYS_* tables are loaded and then their table ids are changed. Some SYS tables like SYS_ZIP_DICT, VIRTUAL can have ids > 1024 (say 1028). When changing table_ids of SYS_FIELDS, from 4 to 1028, it will conflict with the table_id of those existing SYS_ZIP_DICT/VIRTUAL which haven't been shifted by 1024 yet and currently loaded with 1028. Hence, we change ids of these SYS tables in reverse order.

### CODE REVIEW

File: storage/innobase/dict/dict0dict.cc [hunk 1]
```diff
@@ -5841,7 +5841,7 @@ an earlier upgrade. This will update the table_id by adding DICT_MAX_DD_TABLES
 void dict_table_change_id_sys_tables() {
   ut_ad(dict_sys_mutex_own());
 
-  for (uint32_t i = 0; i < SYS_NUM_SYSTEM_TABLES; i++) {
+  for (int i = SYS_NUM_SYSTEM_TABLES - 1; i >= 0; i--) {
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:5859]: Can you please check if the problem exists for user tables?
Also, please add a zipped 5.7 datadir and testcase

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dict.cc [hunk 2]
```diff
@@ -1859,16 +1859,28 @@ void dict_table_change_id_in_cache(
...
+  id_fold = ut_fold_ull(new_id);
 
+  /* Look for a table with the same id: error if such exists */
+  {
+    dict_table_t *table2;
+    HASH_SEARCH(id_hash, dict_sys->table_id_hash, id_fold, dict_table_t *,
+                table2, ut_ad(table2->cached), table2->id == new_id);
+    ut_a(table2 == nullptr);
+  }
+
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:1878]: better to move it after deletion of old id and before insertion with new id.
ie before assigning table->id = new_id

File: storage/innobase/dict/dict0dict.cc (hunk 1 above)
Comment (rahulmalik87) [reply storage/innobase/dict/dict0dict.cc:5859]: problem not exists for user tables. as they are loaded after mapping. added the test case

File: mysql-test/t/mysql_57_inplace_upgrade.test [hunk 3]
```diff
@@ -358,6 +358,28 @@ SELECT * FROM test.t3;
...
+
+--copy_file $MYSQLTEST_VARDIR/std_data/upgrade/sys_table_duplicate_57.zip $MYSQL_TMP_DIR/data57.zip
+--file_exists $MYSQL_TMP_DIR/data57.zip
+--exec unzip -qo $MYSQL_TMP_DIR/data57.zip -d $MYSQL_TMP_DIR
+
+--let $restart_parameters = restart: --datadir=$MYSQLD_DATADIR1
+--let $wait_counter= 10000
+--source include/start_mysqld_no_echo.inc
+
+--echo ###########################################################################
```
Comment (satya-bodapati) [inline mysql-test/t/mysql_57_inplace_upgrade.test]: isn't there some user table in 5.7 user table to check content?

File: storage/innobase/dict/dict0dict.cc [hunk 4]
```diff
@@ -1859,16 +1859,28 @@ void dict_table_change_id_in_cache(
...
     table_id_t new_id)   /*!< in: new id to set */
 {
+  ulint id_fold;
+
   ut_ad(table);
   ut_ad(dict_sys_mutex_own());
   ut_ad(table->magic_n == DICT_TABLE_MAGIC_N);
 
-  /* Remove the table from the hash table of id's */
+  id_fold = ut_fold_ull(new_id);
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc]: ulint id_fold =..

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
After fixing the test, please test on Jenkins and submit here

File: storage/innobase/dict/dict0dict.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc:5859]: right, another reason could be that every user table is loaded with +1024 offset. So there is no conflict.
Incase of SYS_tables, all are loaded with existing ids. Then we change ids of SYS tables.

File: mysql-test/t/mysql_57_inplace_upgrade.test (hunk 3 above)
Comment (rahulmalik87) [reply mysql-test/t/mysql_57_inplace_upgrade.test]: I did not add any data in the user_table. just created one.

Comment (satya-bodapati) [reply mysql-test/t/mysql_57_inplace_upgrade.test]: Please add some user tables and some content. Just to know if the upgrade is really successful.

**Comment by satya-bodapati** [comment]
@rahulmalik87 please fix clang-format test failure next time

File: storage/innobase/dict/dict0dict.cc (hunk 2 above)
Comment (rahulmalik87) [reply storage/innobase/dict/dict0dict.cc:1878]: fixed

File: mysql-test/t/mysql_57_inplace_upgrade.test (hunk 3 above)
Comment (rahulmalik87) [reply mysql-test/t/mysql_57_inplace_upgrade.test]: added

File: storage/innobase/dict/dict0dict.cc (hunk 4 above)
Comment (rahulmalik87) [reply storage/innobase/dict/dict0dict.cc]: fixed

**Comment by satya-bodapati** [comment]
@rahulmalik87 Jenkins link plz

**Comment by satya-bodapati** [comment]
@rahulmalik87 also please squash to single commit

**Comment by rahulmalik87** [comment]
> @rahulmalik87 Jenkins link plz

https://github.com/login/oauth/authorize?client_id=2e8d1d95e73e1e8b7ba9&scope=read:org,user:email&state=_WNC_y9n9QJI0LEzfyo4_CxXxe_

**Comment by satya-bodapati** [comment]
> > @rahulmalik87 Jenkins link plz
> 
> https://github.com/login/oauth/authorize?client_id=2e8d1d95e73e1e8b7ba9&scope=read:org,user:email&state=_WNC_y9n9QJI0LEzfyo4_CxXxe_

@rahulmalik87 This seems to be wrong link. Just redirects to main ps80 page

File: storage/innobase/dict/dict0dict.cc [hunk 5]
```diff
@@ -5841,7 +5850,7 @@ an earlier upgrade. This will update the table_id by adding DICT_MAX_DD_TABLES
 void dict_table_change_id_sys_tables() {
   ut_ad(dict_sys_mutex_own());
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:5852]: I feel we should add comment about this reverse order requirement.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dict.cc (hunk 5 above)
Comment (rahulmalik87) [reply storage/innobase/dict/dict0dict.cc:5852]: > > > @rahulmalik87 Jenkins link plz
> > 
> > 
> > https://github.com/login/oauth/authorize?client_id=2e8d1d95e73e1e8b7ba9&scope=read:org,user:email&state=_WNC_y9n9QJI0LEzfyo4_CxXxe_
> 
> @rahulmalik87 This seems to be wrong link. Just redirects to main ps80 page

https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1176/

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
220228 14:35:41 [  2%] main.mysql_57_inplace_upgrade             [ skipped ]  Test needs 'big-test' or 'only-big-test' option.

Please run only big-test and paste the new Jenkins link here

**Comment by rahulmalik87** [comment]
> 220228 14:35:41 [ 2%] main.mysql_57_inplace_upgrade [ skipped ] Test needs 'big-test' or 'only-big-test' option.
> 
> Please run only big-test and paste the new Jenkins link here

@satya-bodapati Test is with --only-big-test and successful on debug release https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1186/

**Comment by satya-bodapati** [comment]
> > 220228 14:35:41 [ 2%] main.mysql_57_inplace_upgrade [ skipped ] Test needs 'big-test' or 'only-big-test' option.
> > Please run only big-test and paste the new Jenkins link here
> 
> @satya-bodapati Test is with --only-big-test and successful on debug release https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1186/

It seems the test is skipped in release. Maybe part of this testcase used a debug feature
Can you please create separate testcase file? Do not make it big-test.

**Comment by rahulmalik87** [comment]
> > > 220228 14:35:41 [ 2%] main.mysql_57_inplace_upgrade [ skipped ] Test needs 'big-test' or 'only-big-test' option.
> > > Please run only big-test and paste the new Jenkins link here
> > 
> > 
> > @satya-bodapati Test is with --only-big-test and successful on debug release https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1186/
> 
> It seems the test is skipped in release. Maybe part of this testcase used a debug feature Can you please create separate testcase file? Do not make it big-test.

https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1187

File: mysql-test/t/mysql_57_sys_upgrade.test [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+--echo # Stop the default mtr server
+--echo ###########################################################################
+
+--echo # Stop DB server which was created by MTR default
+--let $shutdown_server_timeout = 300
+--source include/shutdown_mysqld.inc
+
+
+--echo ###########################################################################
+--echo # PS#8050: PS 8.0 upgrade (from PS 5.7) crashes with Assertion failure: dict0dict.cc:1236:table2 == nullptr
```
Comment (satya-bodapati) [inline mysql-test/t/mysql_57_sys_upgrade.test]: use PS-XXXX format instead of PS#XXXX

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
looks good. Please fix one minor comment and proceed with squash

File: mysql-test/t/mysql_57_sys_upgrade.test (hunk 6 above)
Comment (rahulmalik87) [reply mysql-test/t/mysql_57_sys_upgrade.test]: fixed. Also test is successful on local with PS- changes

xtrabackup_prepare_func - Binaries are debug compiled
Using 'all' suites
Collecting tests
Checking leftover processes
Removing old var directory
 - WARNING: Using the 'mysql-test/var' symlink
Creating var directory '/home/rahul.malik/MySQL/src/p8/bld/mysql-test/var'
Installing system database
Using parallel: 1

==============================================================================
                  TEST NAME                       RESULT  TIME (ms) COMMENT
------------------------------------------------------------------------------
[ 50%] main.mysql_57_sys_upgrade                 [ pass ]  56919
[100%] shutdown_report                           [ pass ]
------------------------------------------------------------------------------
The servers were restarted 0 times
The servers were reinitialized 0 times
Spent 56.919 of 79 seconds executing testcases

Completed: All 2 tests were successful.

**Review by satya-bodapati — APPROVED** [review]
LGTM!

**Pull request merged**